gnovm/pkg/gnolang/preprocess.go

```diff
@@ -2306,147 +2258,13 @@ func preprocess1(store Store, ctx BlockNode, n Node) Node {
-				} else { // general case
-					for _, v := range n.Values {
-						if cx, ok := v.(*CallExpr); ok {
-							tt, ok := evalStaticTypeOfRaw(store, last, cx).(*tupleType)
-							if ok && len(tt.Elts) != 1 {
-								panic(fmt.Sprintf("multiple-value %s (value of type %s) in single-value context", cx.Func.String(), tt.Elts))
-							}
-						}
-					}
-					// evaluate types and convert consts.
-					if n.Type != nil {
-						// only a single type can be specified.
-						nt := evalStaticType(store, last, n.Type)
-						for i := 0; i < numNames; i++ {
-							sts[i] = nt
-						}
-						// convert if const to nt.
-						for i := range n.Values {
-							checkOrConvertType(store, last, &n.Values[i], nt, false)
-						}
-					} else if n.Const {
-						// derive static type from values.
-						for i, vx := range n.Values {
-							vt := evalStaticTypeOf(store, last, vx)
-							sts[i] = vt
-						}
-					} else { // T is nil, n not const
-						// convert n.Value to default type.
-						for i, vx := range n.Values {
-							if cx, ok := vx.(*ConstExpr); ok {
-								convertConst(store, last, cx, nil)
-								// convertIfConst(store, last, vx)
-							} else {
-								checkOrConvertType(store, last, &vx, nil, false)
-							}
-							vt := evalStaticTypeOf(store, last, vx)
-							sts[i] = vt
-						}
-					}
-					// evaluate typed value for static definition.
-					for i := range n.NameExprs {
-						// consider value if specified.
-						if len(n.Values) > 0 {
-							vx := n.Values[i]
-							if cx, ok := vx.(*ConstExpr); ok &&
-								!cx.TypedValue.IsUndefined() {
-								if n.Const {
-									// const _ = <const_expr>: static block should contain value
-									tvs[i] = cx.TypedValue
-								} else {
-									// var _ = <const_expr>: static block should NOT contain value
-									tvs[i] = anyValue(cx.TypedValue.T)
-								}
-								continue
-							}
-						}
-						// for var decls of non-const expr.
-						st := sts[i]
-						tvs[i] = anyValue(st)
-					}
-				}
-				// define.
-				if fn, ok := last.(*FileNode); ok {
-					pn := fn.GetParentNode(nil).(*PackageNode)
-					for i := 0; i < numNames; i++ {
-						nx := &n.NameExprs[i]
-						if nx.Name == blankIdentifier {
-							nx.Path = NewValuePathBlock(0, 0, blankIdentifier)
-						} else {
-							pn.Define2(n.Const, nx.Name, sts[i], tvs[i])
-							nx.Path = last.GetPathForName(nil, nx.Name)
-						}
-					}
-				} else {
-					for i := 0; i < numNames; i++ {
-						nx := &n.NameExprs[i]
-						if nx.Name == blankIdentifier {
-							nx.Path = NewValuePathBlock(0, 0, blankIdentifier)
-						} else {
-							last.Define2(n.Const, nx.Name, sts[i], tvs[i])
-							nx.Path = last.GetPathForName(nil, nx.Name)
-						}
-					}
+				if len(n.Values) > 1 && len(n.NameExprs) != len(n.Values) {
+					panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %d value(s)", len(n.NameExprs), len(n.Values)))
```

Comment:
Sorry my last request. What do you think about moving this check into the `defineOrDecl` function?

Reply:
basically I think these check should be in AssertCompatible(please consider together with my previous comments @hthieu1110 ). but I'll resolve this since it's done.